ethcore/src/client/client.rs

```diff
@@ -1327,37 +1327,62 @@ impl BlockChainReset for Client {
 	fn reset(&self, num: u32) -> Result<(), String> {
 		if num as u64 > self.pruning_history() {
 			return Err("Attempting to reset to block with pruned state".into())
+		} else if num == 0 {
+			return Err("invalid number of blocks to reset".into())
 		}
 
-		let (blocks_to_delete, best_block_hash) = self.chain.read()
-			.block_headers_from_best_block(num)
-			.ok_or("Attempted to reset past genesis block")?;
+		let mut blocks_to_delete = Vec::with_capacity(num as usize);
+		let mut best_block_hash = self.chain.read().best_block_hash();
 
-		let mut db_transaction = DBTransaction::with_capacity((num + 1) as usize);
-
-		for hash in &blocks_to_delete {
-			db_transaction.delete(::db::COL_HEADERS, &hash.hash());
-			db_transaction.delete(::db::COL_BODIES, &hash.hash());
-			db_transaction.delete(::db::COL_EXTRA, &hash.hash());
-			Writable::delete::<H256, BlockNumberKey>
-				(&mut db_transaction, ::db::COL_EXTRA, &hash.number());
+		for _ in 0..=num {
+			let current_header = self.chain.read().block_header_data(&best_block_hash)
+				.ok_or("Attempted to reset past genesis block")?;
+			best_block_hash = current_header.parent_hash();
+			blocks_to_delete.push((current_header.number(), current_header.hash()));
 		}
 
+		let hashes = blocks_to_delete.iter().map(|(_, hash)| hash).collect::<Vec<_>>();
+		info!("Deleting block hashes {}",
+			  Colour::Red
+				  .bold()
+				  .paint(format!("{:#?}", hashes))
+		);
+
+		let mut db_transaction = DBTransaction::with_capacity(blocks_to_delete.len());
+		let (_, best_block_hash) = blocks_to_delete.first()
+			.ok_or("num is > 0; blocks can't be empty; qed")?;
+
+		for (number, hash) in &blocks_to_delete {
+			db_transaction.delete(::db::COL_HEADERS, &*hash);
+			db_transaction.delete(::db::COL_BODIES, &*hash);
+			Writable::delete::<BlockDetails, H264>(&mut db_transaction, ::db::COL_EXTRA, &*hash);
+			Writable::delete::<H256, BlockNumberKey>
+				(&mut db_transaction, ::db::COL_EXTRA, &*number);
+		}
+
+		let mut best_block_details = Readable::read::<BlockDetails, H264>(
+			&**self.db.read().key_value(),
+			::db::COL_EXTRA,
+			&*best_block_hash
+		).ok_or("block was previously imported; best_block_details should exist; qed")?;
```

Comment:
can this block be removed in the meantime?

Reply:
don't understand the question.

Reply:
We do schedule it for deletion in `db_transaction` above, but it''s not yet comitted, so the block indeed should be there.
However this is an `Err` not expect, so a proof is not even needed in that case.

Reply:
Sorry I meant if the block could be deleted from a separate thread/process. Looks like in this case we're using rocksdb, which doesn't allow read/write from multiple processes and this method is only used in cli, so this is the only thread doing mutation, right? But if change our db backend to lmdb, this is something to keep in mind.